python/ray/serve/tests/fault_tolerance_tests/test_controller_recovery.py

```diff
@@ -135,7 +133,8 @@ def make_nonblocking_calls(expected, expect_blocking=False):
         blocking = []
         responses = defaultdict(set)
         start = time.time()
-        while time.time() - start < 30:
+        timeout_value = 1000 if sys.platform == "win32" else 30
```

Comment:
This limit might seem too much. However, most of the times this test works fine with `timeout_value = 500` (but fails rarely). So, I thought to keep the limit as `1000` to cover every random case when the test might fail in Windows (sometimes it can be too slow).

Reply:
1000s timeout is typically not acceptable in CI system because it take sway too long. Also this test itself is probably put under bazel which put a time limit 
https://github.com/ray-project/ray/blob/master/python/ray/serve/BUILD

300s is the longest acceptable test timeout in CI https://docs.bazel.build/versions/main/test-encyclopedia.html#role-of-the-test-runner. 

I guess another question is _why_ is it so close on windows?

Reply:
Its highly variable on Windows. I can look for the most time consuming block though. May be we can find something worth fixing?

Reply:
I am looking into this.

Reply:
Does the following make sense,

```python
    while time.time() - start < timeout_value:
            refs = [call.remote(block=False) for _ in range(10)]
            _start = time.time()
            ready, not_ready = ray.wait(refs, timeout=5)
            _end = time.time()
            for ref in ready:
                val, pid = ray.get(ref)
                responses[val].add(pid)
            for ref in not_ready:
                blocking.extend(not_ready)
            
            print("ray.wait time: {}".format(_end - _start), len(ready), len(not_ready), attempt,
                  all(
                    len(responses[val]) == num
                    for val, num in expected.items()), responses)

            attempt += 1
            if (all(
                    len(responses[val]) == num
                    for val, num in expected.items())
                    and (expect_blocking is False or len(blocking) > 0)): 
                break
        else:
            assert False, f"Timed out, responses: {responses}."
```

```python
ray.wait time: 5.078817129135132 0 10 0 False defaultdict(<class 'set'>, {'1': set()})
ray.wait time: 2.4971179962158203 1 9 1 False defaultdict(<class 'set'>, {'1': {'6928'}})
ray.wait time: 0.9665451049804688 1 9 2 False defaultdict(<class 'set'>, {'1': {'6928'}})
ray.wait time: 0.36313629150390625 1 9 3 False defaultdict(<class 'set'>, {'1': {'6928'}})
ray.wait time: 0.3152012825012207 1 9 4 False defaultdict(<class 'set'>, {'1': {'6928'}})
ray.wait time: 0.3322627544403076 1 9 5 False defaultdict(<class 'set'>, {'1': {'6928'}})
ray.wait time: 0.4393119812011719 1 9 6 False defaultdict(<class 'set'>, {'1': {'6928'}})
ray.wait time: 1.0798327922821045 1 9 7 False defaultdict(<class 'set'>, {'1': {'6928'}})
ray.wait time: 1.0322465896606445 1 9 8 True defaultdict(<class 'set'>, {'1': {'6096', '6928'}})
ray.wait time: 1.1395301818847656 1 9 0 True defaultdict(<class 'set'>, {'1': {'6928'}})
```

**Crux** - Basically, `ray.wait` is exiting before the timeout specified even though, `ready != 10`.

Reply:
It looks like `ray.wait()` sets `num_returns=1` by default (see https://docs.ray.io/en/latest/package-ref.html#ray-wait). Can you set `num_returns` to 10 (i.e. `ray.wait(refs, num_returns=10, timeout=5)`, and see if `ray.wait` still exits before `ready == 10`?

Reply:
> It looks like `ray.wait()` sets `num_returns=1` by default

By default, I think it should be equal to the length of list (if a list is passed) otherwise 1. Anyways, not a critical issue to deal with.